Fee Estimation via Fee rate Forecasters

ismaelsadeeq opened the pull request:
This PR is the complete implementation of #30392 it aim to improve Bitcoin Core Fee Estimation by:

- **Reducing Overestimation**: Address and mitigate the overestimation issues present in the current `BlockPolicyEstimator`. This issue has been documented and acknowledged by various sources:
  - [See Delving bitcoin posts](https://delvingbitcoin.org/t/mempool-based-fee-estimation-on-bitcoin-core/703/8?u=ismaelsadeeq)
  - [Rene-pickhard Research](https://delvingbitcoin.org/t/mempool-based-fee-estimation-on-bitcoin-core/703/12?u=ismaelsadeeq)
- **Mempool Awareness**: Enable the fee estimator to be aware of the mempool state, allowing it to react faster and more accurately to rapidly changing fee market conditions, when targeting very short timeframes.
- **Empowering Node Users**: Allow node users to be self-sovereign by using their node's estimates, reducing reliance on third-party fee estimations.
- **Simplifying Strategy Integration**: Simplify the process of adding new fee estimation strategies in the future.

The detailed design document for this PR can be found [here](https://gist.github.com/ismaelsadeeq/bcfa6b1216eef10076665551d9885722). 

Please note that the design is subject to change as we refine the approach.

---

This is a collaborative effort with @willcl-ark and incorporates insights from other contributors.

luke-jr commented:
>Make the fee estimator aware of the state of the mempool, allowing it to respond to changing conditions immediately.

The state of the node's mempool may not accurately reflect the state of others' mempools, and not even its own mempool when the block is found in the future. It isn't a good single source of information. Perhaps it is a good idea to use it as a secondary source, but probably it should only ever adjust fee estimations upward, not down.

willcl-ark commented:
> The state of the node's mempool may not accurately reflect the state of others' mempools, and not even its own mempool when the block is found in the future. It isn't a good single source of information.

Correct. The rationale behind this set of changes can be summed up briefly as follows:

- Add a new standalone modular fee estimation manager to which (many) `Forcasters` can be trivially added or removed (vs modifying `BlockPolicyEstimator`).
- Provide two forcaster implementations which do not exist today ( 1. mempool-based and 2. previous-6-blocks-seen-in-mempool) to demonstrate functionality.

In the present, there is a strong tendency for users of Bitcoin Core to consult external fee estimation services when they want timely (next 1 or 2 blocks) confirmation of transactions, whilst also not overpaying. Examples of these include mempool.space, whatthefee.io, Samourai's [nextblock.is](https://blog.samourai.is/introducing-nextblock-fee-estimator/) (now down), [johoe](https://jochen-hoenicke.de/queue/#BTC,1w,weight), [blockchair](https://blockchair.com/bitcoin) etc., with more popping up every month.

We also analysed/estimated Bitcoin Core users not using in-built estimation in a post [here](https://hackmd.io/@kEyqkad6QderjWKtcBF5Hg/cChallengies-with-estimating-transaction-fees).

In our opinion having users feel the need to use external fee estimation services that they could equally have served to them by their own node, feels sub-optimal.

In addition to this, when users do use the current Bitcoin Core fee estimator, there are often times when they end up overpaying, see delving bitcoin post [Mempool based fee estimation](https://delvingbitcoin.org/t/mempool-based-fee-estimation-on-bitcoin-core/703) and various issues over the years e.g. https://github.com/bitcoin/bitcoin/issues/30009 . This is avoidable.

Work from a student of @renepickhardt [link](https://delvingbitcoin.org/t/mempool-based-fee-estimation-on-bitcoin-core/703/12) demonstrated something we also measured independently -- that bitcoin core's current estimates are often overpaying significantly following fee spikes. This effect can be directly mitigated by using a mempool-based estimation.

>  Perhaps it is a good idea to use it as a secondary source, but probably it should only ever adjust fee estimations upward, not down.

In this changeset we take the approach of using the lowest result from all "confident" `Forcasters`. The rationale is that we expect users wanting fast confirmation to have RBF enabled, allowing them to bump fees if we still undershoot.

Comments from @harding [link](https://delvingbitcoin.org/t/mempool-based-fee-estimation-on-bitcoin-core/703/6) explained that it may be possible for miners to _artificially_ increase a strictly-mempool-based fee-rate. By taking the lower (confident) result from `n` `Forcasters`, we attempt to protect against this attack (and others like it), at the potential cost of having to RBF.

We do plan to add additional sanity checks to the mempool-based Forcaster as described in https://github.com/bitcoin/bitcoin/issues/27995, but these are not yet implemented. In any case, even without these additional checks we have been seeing much-improved short time-scale estimations from Bitcoin Core.

willcl-ark committed:
===== End of FeeRateForecasterMan commits =====

willcl-ark committed:
===== End of RPC commits  =====

ismaelsadeeq commented:
Rebased and added newly introduced files to cmakelist file instead makefile

murchandamus commented:
Concept ACK! Thank you for working on this.

vasild commented:
Concept ACK

> Allow node users to be self-sovereign by using their node's estimates, reducing reliance on third-party fee estimations.

> having users feel the need to use external fee estimation services that they could equally have served to them by their own node, feels sub-optimal

Indeed, I do that :face_with_head_bandage:

ismaelsadeeq committed:
fees: add `MemPoolForecaster` class

- The mempool forecaster uses the unconfirmed transactions in the mempool
  to generate a fee rate that a package should have for it to confirm as soon as possible.

Co-authored-by: willcl-ark <will@256k1.dev>

ismaelsadeeq committed:
test: add mempool forecaster unit test

ismaelsadeeq reviewed (commented):
Thanks for your conceptual review, @murchandamus and @vasild. There's also another review by @remyers in the tracking issue: https://github.com/bitcoin/bitcoin/issues/30392#issuecomment-2410695144.

I have updated this PR based on feedback from #30391 and in-person conversations:

1. I am now using the newly introduced `FeeFrac` datatype, which does not lose precision like `CFeeRate`.
2. Updated the `CalculatePercentile` function to return a monotonically decreasing estimate for high and low priority. The previous approach took the fee rate of the package at the exact percentile, which could be high due to the ancestors of the percentile package included previously in a sibling package. 

Relevant commit: 2ac86a8fd7fe108078511ef2f6ad77e315591a2b. 
I've also added a test for this.

3. Added a functional test for the `estimatefee` RPC behavior (commit ac8f2caf050e73b038a8e03a51dfe69e1ea7aa19).

4. Removed the tracing commit after merging #26593, as it changes the tracing structure, which I haven't reviewed yet. The tracing commit was relevant for collecting node's estimation data for analysis, it can be added later on.


This PR has now received a conceptual review from a few contributors, so we can proceed to code and approach review.

ismaelsadeeq committed:
fees: add `forecastTypeToString` method

- This method converts a ForecastType enum to its
  string representation.

ismaelsadeeq committed:
fees: make block_policy_estimator a forecaster

ismaelsadeeq committed:
fees: add `CalculatePercentiles` function

- The CalculatePercentiles function, given
  a vector of feerates in the order they were added
  to the block, will return the 25th, 50th, 75th,
  and 95th percentile feerates.

- Also add a unit test for this function.

ismaelsadeeq committed:
fees: add `ForecastFeeRateFromForecasters` method

- Polls all registered forecasters and selects the lowest fee rate.

ismaelsadeeq committed:
fees: add block policy estimator to forecaster manager

- This changes `CBlockPolicyEstimator` to a shared pointer
  this gives us three advantages.
   - Registering to validation interface using shared pointer
   - Scheduling block policy estimator flushes using shared pointer
   - Registering block policy estimator to forecaster_man

ismaelsadeeq committed:
test: add `estimatesmartfee` new functionality

ismaelsadeeq committed:
fees: introduce lock to forecaster manager and sanity check

- Also add functional test

ismaelsadeeq committed:
==== End Of Sanity Check ====

yancyribbens commented on src/rpc/fees.cpp:
```diff
@@ -94,6 +96,76 @@ static RPCHelpMan estimatesmartfee()
+{
+    return RPCHelpMan{
+        "estimatefee",
+        "\nEstimates the approximate fee per kilobyte needed for a transaction to begin\n"
```
This is confusing I think.  It's titled `estimatefee` however a fee is usually a bitcoin amount.  A feerate however is a bitcoin amount per size.

yancyribbens replied:
should this be "estimatefeerate"?

ismaelsadeeq replied:
I will update this here #31664

ismaelsadeeq replied:
The new RPC is gone; I've integrated this new feature into `estimatesmartfee` and maintained backward compatibility for users who want only block policy, using a flag.

yancyribbens commented on src/test/mempoolforecaster_tests.cpp:
```diff
@@ -0,0 +1,109 @@
+        BOOST_CHECK(fee_estimate.Empty());
+        BOOST_CHECK(*fee_estimate.GetError() == strprintf("Confirmation target %s exceeds the maximum limit of %s. mempool conditions might change, "
+                                                          "making forecasts above %s block may be unreliable",
+                                                          MEMPOOL_FORECAST_MAX_TARGET + 1, MEMPOOL_FORECAST_MAX_TARGET, MEMPOOL_FORECAST_MAX_TARGET));
```
shouldn't the first positional argument be `conf_target` ?

ismaelsadeeq replied:
No, `conf_target` is a struct can not be passed to `strprintf`

yancyribbens replied:
Ok, it's just that you are using `MEMPOOL_FORECAST_MAX_TARGET + 1` as the positional argument for `Confirmation target %s` which doesn't seem right, but maybe I'm mistaken.

yancyribbens commented on src/test/mempoolforecaster_tests.cpp:
```diff
@@ -0,0 +1,109 @@
+    // Test when there are not enough mempool transactions to get an accurate estimate
+    {
+        std::vector<CTransactionRef> highfee_txs;
+        // Add transactions with high_fee fee until mempool transactions weight is more than 25th percent of DEFAULT_BLOCK_MAX_WEIGHT
```
```suggestion
        // Add transactions with high_fee until mempool transactions weight is more than 25th percent of DEFAULT_BLOCK_MAX_WEIGHT
```

yancyribbens replied:
> than 25th percent of DEFAULT_BLOCK_MAX_WEIGHT

Maybe others understand this, but I'm not sure what 25th percent of ... means.  Is that the same as 25% of ..?

ismaelsadeeq replied:
It is` 0.25` * `DEFAULT_BLOCK_MAX_WEIGHT` see https://delvingbitcoin.org/t/mempool-based-fee-estimation-on-bitcoin-core/703/2 where the term was used.

yancyribbens replied:
Oh, you mean to say 25th percentile of  `DEFAULT_BLOCK_MAX_WEIGHT`

ismaelsadeeq committed:
fees: add ForecasterMan class

- Its a module for managing and utilising multiple
  fee rate forecasters to provide fee rate forecast.

- The ForecasterManager class allows for the registration of
  multiple fee rate forecasters.

Co-authored-by: willcl-ark <will@256k1.dev>

ismaelsadeeq committed:
fees: add forecaster structures

i  - Forecaster abstract class is the base class of fee rate forecasters.
     Derived classes must provide concrete implementation
     of the virtual methods.

ii - ForecastResult struct represent the response returned by
     a fee rate forecaster.

iii - ForecastType will be used to identify forecasters.
      Each time a new forecaster is added, a corresponding
      enum value should be added to ForecastType.

Co-authored-by: willcl-ark <will@256k1.dev>

ismaelsadeeq commented:
Rebased

Ferrydh commented:
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

ismaelsadeeq committed:
fees: cache `MemPoolPolicyEstimator` forecasts

- Provide new forecast only when the time delta from previous
  forecast is older than 30 seconds.

- This caching helps avoid the high cost of frequently generating block templates.

Co-authored-by: willcl-ark <will@256k1.dev>

willcl-ark commented:
I restarted my relatively long-running node after building this PR and notice that now I get the following when trying to get any estimate:

```sh
src/core/bitcoin on  pr-30157 [$?] via △ v3.31.6 via 🐍 v3.13.3 via ❄️  impure (nix-shell-env)
❯ /home/will/src/core/bitcoin/build/bin/bitcoin-cli estimatesmartfee 1
{
  "blocks": 0,
  "errors": [
    "Insufficient data (at least 6 blocks needs to be processed after startup)"
  ]
}

src/core/bitcoin on  pr-30157 [$?] via △ v3.31.6 via 🐍 v3.13.3 via ❄️  impure (nix-shell-env)
❯ /home/will/src/core/bitcoin/build/bin/bitcoin-cli estimatesmartfee 3
{
  "blocks": 0,
  "errors": [
    "Insufficient data (at least 6 blocks needs to be processed after startup)"
  ]
}
```

This is not the case when I restart v29.0:

```sh
src/core/bitcoin on  pr-30157 [$?] via △ v3.31.6 via 🐍 v3.13.3 via ❄️  impure (nix-shell-env)
❯ bitcoind
Bitcoin Core starting

src/core/bitcoin on  pr-30157 [$?] via △ v3.31.6 via 🐍 v3.13.3 via ❄️  impure (nix-shell-env)
❯ /home/will/src/core/bitcoin/build/bin/bitcoin-cli estimatesmartfee 2
{
  "feerate": 0.00005081,
  "blocks": 2
}
```

It's able to return a fee estimate to me immediately, as it has saved estimation data from the previous run. I don't think we should break this behaviour here; waiting ~ an hour to get a new estimate after a restart is not a pleasant UX.

At the very least we should not nerf blockpolicy estimations as part of this change. Although, I think thought that we should be smart enough to know if we have a ~ up-to-date mempool and use that data if we do, to return estimates via the mempool-based estimator. If we need, perhaps we can return a warning about the age of the mempool when returning e.g. a next-block estimate, although with RBF prevalance now I'm not sure even that is necessary?

willcl-ark commented:
> Yep, the node needs to see six blocks mined before making an estimate. Were you able to see an estimate after six blocks were mined?

I'm still waiting 😄  but I assume that I will be able to eventually...

When we have both `fee_estimates.dat` and `mempool.dat` saving all the relevant data I don't think we should be introducing a 6 block wait to get fee estimates. My data was ~ 10 seconds out of date after a restart, and now I must wait an hour.

ismaelsadeeq committed:
rpc: update `estimatefeesmartfee` to use forecaster_man

- Given a confirmation target, we use fee estimator module that call all
available fee estimator forcasters and return the lowest fee rate that if
a transaction use will likely confirm in a given confirmation target.

- This also provides backward compatibility

Co-authored-by: willcl-ark <will@256k1.dev>
